### PR TITLE
use mongodb extension instead of mongo extension

### PR DESCRIPTION
**Pull request by sandrokeil**
This is a PR for #33 
- add option to use expireAfterSeconds index, default disabled
- add install mongodb extension on Travis-CI

### CODE REVIEW

**Comment by sandrokeil** [comment]
@weierophinney All tests are green now. :-) Should I rename the class MongoDB to MongoDb?

**Comment by sandrokeil** [comment]
Before merge, take a look at the discussion in the issue #33 for this PR. Maybe there is additional work required.

File: test/SaveHandler/MongoDBTest.php [hunk 1]
```diff
@@ -43,8 +44,10 @@ class MongoDBTest extends \PHPUnit_Framework_TestCase
      */
     public function setUp()
     {
-        if (!extension_loaded('mongo')) {
-            $this->markTestSkipped('Zend\Session\SaveHandler\MongoDB tests are not enabled due to missing Mongo extension');
+        if (!extension_loaded('mongodb')) {
```
Comment (Maks3w) [inline test/SaveHandler/MongoDBTest.php]: I think we should go rid of this check and use `@requires` annotations  https://phpunit.de/manual/current/en/incomplete-and-skipped-tests.html#incomplete-and-skipped-tests.requires.tables.api

File: src/SaveHandler/MongoDB.php [hunk 2]
```diff
@@ -10,62 +10,60 @@
...
+use MongoDB\BSON\UTCDatetime;
+use MongoDB\Client as MongoClient;
+use MongoDB\Collection as MongoCollection;
 use Zend\Session\Exception\InvalidArgumentException;
 
 /**
  * MongoDB session save handler
  */
-class MongoDB implements SaveHandlerInterface
+final class MongoDB implements SaveHandlerInterface
```
Comment (Maks3w) [inline src/SaveHandler/MongoDB.php]: Please revert.

File: src/SaveHandler/MongoDB.php [hunk 3]
```diff
@@ -10,62 +10,60 @@
...
+     */
+    private $mongoClient;
+
+    /**
      * MongoCollection instance
      *
      * @var MongoCollection
      */
-    protected $mongoCollection;
+    private $mongoCollection;
```
Comment (Maks3w) [inline src/SaveHandler/MongoDB.php]: Please revert

File: src/SaveHandler/MongoDB.php [hunk 4]
```diff
@@ -10,62 +10,60 @@
...
 
     /**
      * Session name
      *
      * @var string
      */
-    protected $sessionName;
+    private $sessionName;
```
Comment (Maks3w) [inline src/SaveHandler/MongoDB.php]: Please revert.

File: src/SaveHandler/MongoDB.php [hunk 5]
```diff
@@ -10,62 +10,60 @@
...
 
     /**
      * Session lifetime
      *
      * @var int
      */
-    protected $lifetime;
+    private $lifetime;
```
Comment (Maks3w) [inline src/SaveHandler/MongoDB.php]: Please revert.

File: src/SaveHandler/MongoDB.php [hunk 6]
```diff
@@ -10,62 +10,60 @@
...
 
     /**
      * MongoDB session save handler options
      * @var MongoDBOptions
      */
-    protected $options;
+    private $options;
```
Comment (Maks3w) [inline src/SaveHandler/MongoDB.php]: Please revert.

File: src/SaveHandler/MongoDBOptions.php [hunk 7]
```diff
@@ -68,6 +68,12 @@ class MongoDBOptions extends AbstractOptions
      */
     protected $modifiedField = 'modified';
 
+    /**
+     * Use expireAfterSeconds index
+     *
+     * @var bool
+     */
+    protected $useExpireAfterSecondsIndex = false;
```
Comment (Maks3w) [inline src/SaveHandler/MongoDBOptions.php:76]: It seems you are adding a feature. Please do it in a different PR

Comment (sandrokeil) [inline src/SaveHandler/MongoDBOptions.php:76]: Yes, as mentioned above. I will do that.

File: src/SaveHandler/MongoDB.php (hunk 2 above)
Comment (sandrokeil) [inline src/SaveHandler/MongoDB.php]: Sure that I should remove final and private definitions? Thought that's common if only an interface is implemented. Why do you want to allow inheritance of this class in the new major version (v3)?

Comment (Maks3w) [inline src/SaveHandler/MongoDB.php]: Changes in the inheritance are not backward compatible.

File: src/SaveHandler/MongoDBOptions.php (hunk 7 above)
Comment (weierophinney) [inline src/SaveHandler/MongoDBOptions.php:76]: I'd argue differently. This PR:
- is swapping out the underlying extension used
- gaining the ability to expose features previously unavailable (due to extension limitations)

As such, it's already targeted at a new minor version. Pushing the feature into a later PR creates more work for @sandrokeil, when we can easily evaluate it in this same pull request.

File: src/SaveHandler/MongoDB.php (hunk 2 above)
Comment (weierophinney) [inline src/SaveHandler/MongoDB.php]: Agreed with @Maks3w on this one. If this were a new class, final/private would be fine, but for existing classes, we need to keep visibility the same.

We should definitely revisit for v3.

**Comment by sandrokeil** [comment]
Thanks for clarification. I've updated the PR to restore BC for version 2.7. One problem is how to ensure that people have to use the _mongodb_ extension. As mentioned by @Maks3w there is no Composer constraint.

**Comment by weierophinney** [comment]
@sandrokeil :+1: 

``` json
"require": {
    "ext/mongodb": "*"
}
```

_will_ fail if the extension is not present (however, it will _not_ fail if you use the `--ignore-platform-reqs` flag, but if you do that, you're setting yourself up for problems anyways).

**Comment by weierophinney** [comment]
Okay, just finally figured out what you meant by "there is no Composer constraint": the `composer.json` does not contain one currently _because this is optional functionality_.

As such, we have three options:
- indicate in the `suggest` section that users should install `ext/mongodb` in order to use the MongoDB save handler. This, of course, breaks functionality for users who were previously using `ext/mongo` when they upgrade.
- do the above, but also raise an exception in the `MongoDB` save handler if `ext/mongodb` is not loaded; the exception message should indicate what extension is needed, and could also indicate how to downgrade if the exception has occurred since an upgrade.
- release the updated functionality _as a separate package_, and keep the existing functionality. We would then document to users that if they want the updated code or want to use `ext/mongodb`, they should also install the new package. If we go this route, we should also mark the current save handler as deprecated, and likely raise an `E_USER_DEPRECATED` from its constructor to signal users to start using the new package. This would allow adding the extension constraint to the new package's `composer.json`.

Pinging @ezimuel and @zendframework/community-review-team for feedback.

**Comment by weierophinney** [comment]
Just a quick note: separation to a new package is the route that @marc-mabe is planning for storage adapters in zend-cache, for exactly this same reason.

**Comment by sandrokeil** [comment]
Yes, separation for each special save handler, which relies on dependencies, in an extra package is the best choice. I'm for option 3 and waiting for the feedback.

File: src/SaveHandler/MongoDB.php [hunk 8]
```diff
@@ -10,10 +10,10 @@
 
 namespace Zend\Session\SaveHandler;
 
-use Mongo;
-use MongoClient;
-use MongoCollection;
-use MongoDate;
+use MongoDB\BSON\Binary;
+use MongoDB\BSON\UTCDatetime;
+use MongoDB\Client as MongoClient;
```
Comment (prolic) [inline src/SaveHandler/MongoDB.php]: This class does not exist.

File: src/SaveHandler/MongoDB.php [hunk 9]
```diff
@@ -10,10 +10,10 @@
...
+use MongoDB\Collection as MongoCollection;
```
Comment (prolic) [inline src/SaveHandler/MongoDB.php]: This class does not exist.

File: src/SaveHandler/MongoDB.php (hunk 8 above)
Comment (prolic) [inline src/SaveHandler/MongoDB.php:15]: It's only available in mongo php library. We don't need this. We can do everything with the raw extension here, imho.

File: src/SaveHandler/MongoDB.php (hunk 9 above)
Comment (prolic) [inline src/SaveHandler/MongoDB.php:16]: It's only available in mongo php library. We don't need this. We can do everything with the raw extension here, imho.

**Comment by sandrokeil** [comment]
@prolic Please take a look at [this comment](https://github.com/zendframework/zend-session/issues/33#issuecomment-203047117). But yes, we can also use the raw driver instead of the library.

File: src/SaveHandler/MongoDB.php (hunk 8 above)
Comment (weierophinney) [inline src/SaveHandler/MongoDB.php:15]: We can, but it requires far more code and code complexity. The Mongo PHP Library exists to provide the convenience layer that was previously in the extension, but in a way that can be more quickly and easily updated as new features are added and/or bugfixes released. Considering most developers will now be using the extension / PHP library combination if using MongoDB at all on PHP, this is not an unreasonable requirement.
